From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390/time: use stck clock fast for do_account_vtime
Patch-mainline: v3.18-rc4
Git-commit: 1f759bb3a2a0d75ceeeec729b1c66a7f443631ba
References: bnc#925012

Summary:     s390: z13 base performance
Description: Provide the backport of a minimum set of upstream patches
             to optimize the base performance of the IBM z13 machine.

Upstream-Description:

             s390/time: use stck clock fast for do_account_vtime

             The last high frequency call site of the STCK instruction is
             do_account_vtime. Replace it with the faster STCKF instruction.

             Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/kernel/vtime.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/arch/s390/kernel/vtime.c
+++ b/arch/s390/kernel/vtime.c
@@ -62,7 +62,13 @@ static void do_account_vtime(struct task
 	timer = S390_lowcore.last_update_timer;
 	clock = S390_lowcore.last_update_clock;
 	asm volatile ("  STPT %0\n"    /* Store current cpu timer value */
+#if defined(CONFIG_MARCH_Z9_109) || \
+    defined(CONFIG_MARCH_Z10) || \
+    defined(CONFIG_MARCH_Z196)
+		      "  STCKF %1"     /* Store current tod clock value */
+#else
 		      "  STCK %1"      /* Store current tod clock value */
+#endif
 		      : "=m" (S390_lowcore.last_update_timer),
 		        "=m" (S390_lowcore.last_update_clock) );
 	S390_lowcore.system_timer += timer - S390_lowcore.last_update_timer;
